src/wp-includes/class-wp-modules.php

```diff
@@ -0,0 +1,251 @@
+<?php
+/**
+ * Modules API: WP_Modules class.
+ *
+ * Native support for ES Modules and Import Maps.
+ *
+ * @package WordPress
+ * @subpackage Modules
+ */
+
+/**
+ * Core class used to register modules.
+ *
+ * @since 6.5.0
+ */
+class WP_Modules {
+	/**
+	 * Holds the registered modules, keyed by module identifier.
+	 *
+	 * @since 6.5.0
+	 * @var array
+	 */
+	private $registered = array();
+
+	/**
+	 * Holds the module identifiers that were enqueued before registered.
+	 *
+	 * @since 6.5.0
+	 * @var array
+	 */
+	private $enqueued_before_registered = array();
+
+	/**
+	 * Registers the module if no module with that module identifier has already
+	 * been registered.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string            $module_identifier The identifier of the module. Should be unique. It will be used in the
+	 *                                             final import map.
+	 * @param string            $src               Full URL of the module, or path of the script relative to the WordPress
+	 *                                             root directory.
+	 * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this
+	 *                                             module. The dependencies can be strings or arrays. If they are arrays,
+	 *                                             they need an `id` key with the module identifier, and can contain a
+	 *                                             `type` key with either `static` or `dynamic`. By default, dependencies
+	 *                                             that don't contain a type are considered static.
+	 * @param string|false|null $version           Optional. String specifying module version number. Defaults to false.
+	 *                                             It is added to the URL as a query string for cache busting purposes. If
+	 *                                             $version is set to false, the version number is the currently installed
+	 *                                             WordPress version. If $version is set to null, no version is added.
+	 */
+	public function register( $module_identifier, $src, $dependencies = array(), $version = false ) {
+		if ( ! isset( $this->registered[ $module_identifier ] ) ) {
+			$deps = array();
+			foreach ( $dependencies as $dependency ) {
+				if ( isset( $dependency['id'] ) ) {
```

Comment:
Either this or in an `else` (see below) at the end would perhaps be more robust to help authors.

```suggestion
				if ( is_array( $dependency ) ) {
					if ( ! isset( $dependency['id'] ) ) {
						_doing_it_wrong( __METHOD__, __( 'Missing required id key in entry among dependencies array.' ), '6.5.0' );
						continue;
					}
```

Reply:
Makes sense, thank you. Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/b3554d61c10ab2010d66178eca87d7dee62a470e.